saleor/checkout/calculations.py

```diff
@@ -437,45 +456,50 @@ def _fetch_checkout_prices_if_expired(
             # tax from the original gross prices.
             _remove_tax(checkout, lines)
 
-    checkout_update_fields = [
-        "voucher_code",
-        "total_net_amount",
-        "total_gross_amount",
-        "subtotal_net_amount",
-        "subtotal_gross_amount",
-        "shipping_price_net_amount",
-        "shipping_price_gross_amount",
-        "undiscounted_base_shipping_price_amount",
-        "shipping_tax_rate",
-        "translated_discount_name",
-        "discount_amount",
-        "discount_name",
-        "currency",
-        "last_change",
-        "price_expiration",
-        "tax_error",
-    ]
-
     checkout.price_expiration = timezone.now() + settings.CHECKOUT_PRICES_TTL
 
-    from .utils import checkout_lines_bulk_update
-
-    with allow_writer():
-        with transaction.atomic():
-            checkout.save(
-                update_fields=checkout_update_fields,
-                using=settings.DATABASE_CONNECTION_DEFAULT_NAME,
-            )
-            checkout_lines_bulk_update(
-                [line_info.line for line_info in lines],
-                [
-                    "total_price_net_amount",
-                    "total_price_gross_amount",
-                    "tax_rate",
-                    "undiscounted_unit_price_amount",
-                    "prior_unit_price_amount",
-                ],
-            )
+    # Check whether the checkout has been modified during the recalculation process. If so, we should skip saving.
+    # The same applies if the checkout has been removed. This is important to avoid overwriting changes made by
+    # the other requests. Skipping the save function does not affect the query response because it returns
+    # the adjusted checkout and line info objects.
+    if not _is_checkout_modified(checkout, database_connection_name):
```

Comment:
IMHO you still have a "slot" for mismatch here:
- (Worker1) _is_checkout_modified -> False 
- (Worker2) price_expired=True
- (Worker1) checkout.save(...)

I think that, you need to call your checker inside the transaction + lock.